src/jdk.hotspot.agent/share/classes/sun/jvm/hotspot/CommandProcessor.java

```diff
@@ -1773,20 +1772,49 @@ public void doit(Tokens t) {
                 sysProps.run();
             }
         },
-        new Command("dumpheap", "dumpheap [filename]", false) {
+        new Command("dumpheap", "dumpheap [gz=<1-9>] [filename]", false) {
             public void doit(Tokens t) {
-                if (t.countTokens() > 1) {
+                int cntTokens = t.countTokens();
+                if (cntTokens > 2) {
                     usage();
                 } else {
                     JMap jmap = new JMap();
-                    String filename;
-                    if (t.countTokens() == 1) {
+                    String filename = "heap.bin";
+                    int gzlevel = 0;
+                    /*
+                     * Possible command:
+                     *     dumpheap gz=1 file;
+                     *     dumpheap gz=1;
+                     *     dumpheap file;
+                     *     dumpheap
+                     *
+                     * Use default filename if cntTokens == 0.
+                     * Handle cases with cntTokens == 1 or 2.
+                     */
+                    if (cntTokens > 2) {
+                        err.println("Too big number of options: " + cntTokens);
+                        usage();
+                        return;
+                    }
+                    if (cntTokens >= 1) { // parse first argument which is "gz=" option
+                        String option  = t.nextToken();
+                        gzlevel = parseHeapDumpCompressionLevel(option);
+                        if (gzlevel <= 0 || gzlevel > 9) {
+                            usage();
+                            return;
+                        }
+                        filename = "heap.bin.gz";
+                    }
```

Comment:
I don't think this section works properly if there is just one argument, and it is the filename, not gz=. Also, there appears to be no test case to catch this bug.

Reply:
Oh yes, I think there should be a check whether the option is "gz=" first. And I will the test case for that together when with the fix of HeapDumpTest.java.